### PR TITLE
Make the translation function fully type-safe

### PR DESCRIPTION
**Pull request by pedrodurek**
With Typescript 4.1, now it is possible to use recursive conditional type and template literal string. This PR aims to make the `t` function fully typed and supporting single and multiple namespaces. It works with `useTranslation (hooks)`, `withTranslation (HoC)`, `Trans Component` and `Translation (render prop)`.

### How to use it
You just have to rely on type augmentation technique to override the default Resources with the resource language object.

**For example:**
```ts
 // src/i18n/config.ts
export const resourcesEN = {
  earnings: {
    header: {
      title: 'Earnings',
      subTitle: 'See your Earnings'
    },
    notes: 'Balance'
  }
};

i18n.use(initReactI18next).init({
  resources: {
    en: resourcesEN,
  },
});

```
Type augmentation
```ts
// @types/react-i18next/index.d.ts
import "react-i18next";
import resourcesEN from "../../src/i18n/config";

declare module "react-i18next" {
  type DefaultResources = typeof resourcesEN;
  interface Resources extends DefaultResources {}
}
```
And that's all!!! it works like magic!
I've prepared a few examples using `useTranslation`.

**Using single and multiple namespaces:**
[https://tsplay.dev/gWozjm](https://tsplay.dev/gWozjm)

**Array case within the resources JSON:**
[https://tsplay.dev/nmq0ZN](https://tsplay.dev/nmq0ZN)

**Without Namespace:**
[https://tsplay.dev/rw21xW](https://tsplay.dev/rw21xW)

Below is the diff of the changes that I made:
[https://gist.github.com/pedrodurek/b23c17bebd9619b0dc0651bc3949a82e/revisions#diff-7aa4473ede4abd9ec099e87fec67fd57afafaf39e05d493ab4533acc38547eb8](https://gist.github.com/pedrodurek/b23c17bebd9619b0dc0651bc3949a82e/revisions#diff-7aa4473ede4abd9ec099e87fec67fd57afafaf39e05d493ab4533acc38547eb8)

Let me know if you want me to change anything, thank you guys! I hope this can help everyone who relies on i18next and typescript.

#### Checklist

- [x] only relevant code is changed (make a diff before you submit the PR)
- [x] run tests `npm run test`
- [ ] tests are included
- [ ] documentation is changed or added

### CODE REVIEW

**Comment by DanielRosenwasser** [comment]
I don't know a ton about this library, and I'll admit I don't really understand the implementation, but before merging I'd encourage a bit of testing on the "worst-case" scenarios. You don't want to launch a ton of work on every keystroke.

Does editing with a lot of fields become slow? Do compilation times increase by a decent amount if you add 1000 fields?

**Comment by pedrodurek** [comment]
Hey @DanielRosenwasser, thank you for taking some time to look at my PR and express your opinion, that's a really good point. I've performed some tests on the project that I've been working on, which has close to 2000 keys under the same namespace (legacy), and other hundreds of keys under different namespaces, and I didn't notice any significant increase in terms of time compilation, memory consumption and CPU. 
To be honest, I got really great results, It brought the keys in a matter of seconds (1-2 sec). Of course, it can vary according to the computer's specs. So, I'm going to create a side project with a similar case and showing some results. In addition, I'm going to perform the same tests in an old computer that I have and post some results here.

**Comment by DanielRosenwasser** [comment]
> To be honest, I got really great results, It brought the keys in a matter of seconds (1-2 sec)

Well it'll be relative. If it's no slower than today, that might be okay; however, we really do aim for less than half a second on completions. Part of me really wishes the speed tests we have in Definitely Typed were more easily runnable to get a sense of regressions here.

**Comment by pedrodurek** [comment]
Hey guys, I've created this repository [https://github.com/pedrodurek/i18n-type-tests](https://github.com/pedrodurek/i18n-type-tests) that contains examples of the "worst-case" scenarios, I encourage everyone who is reviewing this PR to take a look at it.

In a nutshell, the timing to bring the keys considering that ns1 & ns2 have 2548 keys combined takes approximately 0.5~1 sec. Once we import ns3 as well, the total number of keys is 7647 and it's taking 2-3 seconds mostly.

**Comment by mikelfcosta** [comment]
Hey, wondering what's missing for this PR, anything I can help out? Really wanted to use this on my projects with TS 4.1

**Review by adrianolsk — APPROVED** [review]
This looks pretty good, I didn't see any performance decrease even with large objects

**Review by mikelfcosta — APPROVED** [review]
Tested on my side-projects and it handles really well!

**Comment by tigerabrodi** [comment]
If any help or review is needed, I am down to help :tada:

**Comment by pedrodurek** [comment]
Thanks @tigerabrodi, I'd love some help 🤗

**Comment by tigerabrodi** [comment]
@pedrodurek Will block out a good 30 - 60 minutes today, I think I myself can learn a thing or two, thanks for allowing me :two_hearts: :smiling_face_with_three_hearts:

**Comment by jamuhl** [comment]
Just let me know when you think this PR is ready to be merged and rolled out...

**Comment by tigerabrodi** [comment]
As mentioned, I will give it a review later (in 4-5h), really looking forward to contributing and working with you guys on react-i18next :smiling_face_with_three_hearts:

**Review by tigerabrodi — APPROVED** [review]
LGTM.

Just few comments, could also be because of my lack of experience/knowledge.

Overall, I would point out, we should try to avoid `any` type as much as possible, but I bet you already know this :+1: :two_hearts: :tada:

**Comment by pedrodurek** [comment]
Thank you @tigerabrodi, I really appreciate the rapid review! You're completely right, we should always avoid using any, unless  you're using it for manipulation reasons, for example:
Conditional Types
```ts
type Test<T> = T extends any[] ? string : number;
```
or like this, where we're omitting any array from T:

```ts
type OmitArrayProps<T> = Exclude<T, keyof any[]>;
```

This PR aims to make the `t` function from `useTranslation`, `withTranslation`, `Trans Component` and `Translation` hooks/components, so I didn't focus much on the types from the original code (https://github.com/i18next/react-i18next/blob/master/src/index.d.ts). 

Here we can see a comparison of the changes that I've made: https://gist.github.com/pedrodurek/b23c17bebd9619b0dc0651bc3949a82e/revisions#diff-7aa4473ede4abd9ec099e87fec67fd57afafaf39e05d493ab4533acc38547eb8

File: src/ts4.1/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,196 @@
+import i18next, { ReactOptions, i18n, ThirdPartyModule, Resource, TOptions, StringMap } from 'i18next';
+import * as React from 'react';
+
+/**
+ * This interface can be augmented by users to add types to `react-i18next` default resources.
+ */
+export interface Resources {}
+
+type Omit<T, K> = Pick<T, Exclude<keyof T, K>>;
```
[the remark this answers is not in the document]
Comment (pedrodurek) [reply src/ts4.1/index.d.ts:9]: Good catch, this is coming from the original code, but it won't hurt if I remove it.

File: src/ts4.1/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+type ResourcesKey<T = keyof Resources> = [T] extends [never] ? string : T;
+
+export type Namespace = ResourcesKey | ResourcesKey[];
+
+export function setDefaults(options: ReactOptions): void;
+export function getDefaults(): ReactOptions;
+export function setI18n(instance: i18n): void;
+export function getI18n(): i18n;
+export const initReactI18next: ThirdPartyModule;
+export function composeInitialProps(ForComponent: any): (ctx: unknown) => Promise<any>;
```
[the remark this answers is not in the document]
Comment (pedrodurek) [reply src/ts4.1/index.d.ts:20]: Those types are coming from the original code, so I don't know much about the `composeInitialProps` function, but I definitely agree we shouldn't be using any, so we could create an issue to tackle that later.

File: src/ts4.1/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+    key: K,
+    defaultValue?: string,
+    options?: TOptions<TInterpolationMap> | string,
+  ): TFuncReturn<N, K>;
+}
+
+export interface TransProps<N extends Namespace, K extends TFuncKey<N>, E extends Element = HTMLDivElement>
+  extends React.HTMLProps<E> {
+  children?: React.ReactNode;
+  components?: readonly React.ReactNode[] | { [tagName: string]: React.ReactNode };
```
[the remark this answers is not in the document]
Comment (pedrodurek) [reply src/ts4.1/index.d.ts:106]: Again, this is coming from the original code, but this is definitely right, readonly won't have any effect if placed after.

**Comment by tigerabrodi** [comment]
@pedrodurek Hmm, I also just approved it :two_hearts: :tada:. Anyways, great job :clap:, I really look forward to working with you guys :smiley:.

Also, thank you for the explanation, means a lot :pray: :smiling_face_with_three_hearts:.

**Comment by jamuhl** [comment]
@pedrodurek @tigerabrodi so I guess we're ready to merge this on monday 👍

**Comment by tigerabrodi** [comment]
If any improvements or bugs get found in the future, we can just fix them then as @pedrodurek mentioned :tada: :muscle: :raised_hands: :clap:

**Comment by adrai** [comment]
trying to support the community by helping in the gh issues (and perhaps also on stackoverflow) with some sort of first-level assistance would be awesome...

**Comment by tigerabrodi** [comment]
Jupp :raised_hands:, I think I will put my focus on Github issues :tada:.

**Comment by jamuhl** [comment]
published in react-i18next@11.8.0

**Comment by karol-majewski** [comment]
I'm happy to report that it works perfectly on a codebase with multiple namespaces and ~115 translated phrases.

For anyone storing translations as JSON files just like me: your declaration can be as simple as:

```ts
declare module 'react-i18next' {
  interface Resources {
    foo: typeof import('path/to/foo.json');
    bar: typeof import('path/to/bar.json');
  }
}

export {};
```

**Comment by DCzajkowski** [comment]
How do you handle dynamic keys with this change?

Example:
```js
// resources

const resourcesEN = {
  variants: {
    'a': 'text a',
    'b': 'text b',
  },
}

// component

const variant = 'a' as 'a' | 'b'

t(`variants.${variant}`) // fails
```

**Comment by mikelfcosta** [comment]
Glad to see more people testing this out and checking the many opportunities, should we maybe have a section on the documentation for TypeScript 4.1 with all of these different solutions?

**Comment by DCzajkowski** [comment]
@mikelfcosta 

> Glad to see more people testing this out and checking the many opportunities

For us this PR introduced a breaking change on minor version, as after this minor upgrade our code stopped building — every translation key is now type checked, and previously it wasn't 😬

I am now sitting a third hour trying to get to green again 😅

> should we maybe have a section on the documentation for TypeScript 4.1 with all of these different solutions

To be honest this feels like a big missing thing. Using Typescript with this package is definitely not easy and there is a lot of guess work involved. Having some documentation on this would be ideal! If someone starts some work on this, I'd gladly contribute 👍

**Comment by mikelfcosta** [comment]
> For us this PR introduced a breaking change on minor version, as after this minor upgrade our code stopped building — every translation key is now type checked, and previously it wasn't 😬
> 
> I am now sitting a third hour trying to get to green again 😅
> 
I hadn't looked at that point, you're right. I'll take a look and start helping with some documentation.

**Comment by pedrodurek** [comment]
Just for reference https://github.com/i18next/i18next/issues/1504#issuecomment-742727015

**Comment by joelpoloney** [comment]
> How do you handle dynamic keys with this change?
> 
> Example:
> 
> ```js
> // resources
> 
> const resourcesEN = {
>   variants: {
>     'a': 'text a',
>     'b': 'text b',
>   },
> }
> 
> // component
> 
> const variant = 'a' as 'a' | 'b'
> 
> t(`variants.${variant}`) // fails
> ```
> 
> **EDIT:**
> 
> I have found a solution 😅
> 
> The execution should be:
> 
> ```js
> t(`variants.${variant}` as const) // passes
> ```

Is there a better solution than casting to const everywhere?

**Comment by pedrodurek** [comment]
Hey @joelpoloney, unfortunately that's the only way. That's a typescript behaviour ("deficiency"), when we concatenate strings, typescript automatically infers the type as string, and in order to get the literal values, we must assert the value. Take a look at this [example](https://www.typescriptlang.org/play?#code/MYewdgzgLgBAZgVzMGBeGAKAhgJwOYBcMA5AA5ZQAWAdFCNRCALYCmVAlmHtTixAgCMm7KLREAbFsRgAfEuSq16jVhy4NBw0VAlSAlGgB8MAN4BfAFBXQkWOwgBlTSKgsAJmnhZxEFgG4LG2gYdgBGAA4wACEsX097JyEXdxgAfnkKGjoGZjZKTm5efiSoaSIyTKUc1Xz1Yq1iAMDwYJ0oSU9EZAwAAwASEzDImN8zMXaWHr0-GAB6WZgAERA+GDAQWAB3EBwAa2bbGDbJACZOpGBegaHo2JYx48mYWJggqGmrK3mYAEk4GAAniAEDBSOIsMAWEdKFDgAgcIwcDABCxxCBNgAaQHA4jicQwXxQqgUI4A0h8YA4dikLbsPEhMBwFhIqhEslQl7QKlcA7BFgADywTDBLFCaH6gwit1G40kPSaFm+f2xILBEKJMNe8MRyNR6KxQIQuPxhOhJKg7IglOptPpvCg8LA0I5pFIOBAbvYFCh4hEzO8MAAbt4EHxebABUKRWdUBKbiN7rKni83n4gA).

Screen Shot 2020-12-29 at 6 14 30 PM
Screen Shot 2020-12-29 at 6 14 35 PM

**Comment by joelpoloney** [comment]
@pedrodurek ah interesting. I'd have thought TypeScript would be a bit smarter around this. Thank you for the explanation!

**Comment by fiznool** [comment]
I apologise for the negativity here, as I know how much effort it is to maintain an open source library, and I must say that react-i18next is a pleasure to use. Please keep up the good work.

That being said, I just wanted to add a bit of feedback here, in that upgrading from 11.7 -> 11.8 has broken my codebase, in a way that I'm not sure how to fix without a decent amount of effort. I am probably using the library incorrectly, specifically around the typings, but regardless I wouldn't expect a minor jump to cause breaking changes. IMHO, this change should have been reserved for a major bump. I've added some context below.

---

Here is an excerpt from my codebase:

``` json
// views.json
{
  "reviews": {
    "connectPermissions": {
      "facebook": {
        "heading": "About to Connect Facebook",
        "lead": "In a moment, you might be asked by Facebook if you are happy to give permission to:",
        "permissionsList": [
          {
            "permission": "Show a list of Pages you manage",
            "reason": "So you can select the page you want to connect."
          },
          {
            "permission": "Read content posted on the Page",
            "reason": "Allowing us show you a list of all your ratings."
          }
        ]
      },
      "google": {
        "heading": "About to Connect Google",
        "lead": "In a moment, you'll be asked by Google if you are happy to give permission to:",
        "permissionsList": [
          {
            "permission": "See, edit, create and delete your Google business listings ",
            "reason": "We will only use this access to see the list of your businesses, notify you of new reviews, and allow you to reply to any review you receive."
          }
        ]
      }
    }
  }
}


``` ts
// use-scoped-translation.ts

import { StringMap, TFunctionResult, TOptions } from 'i18next';
import { useTranslation } from 'react-i18next';

/**
 * Produces a 'scoped' translate function, eliminating the need to specify the entire key for a section of strings.
 *
 * To use, ensure the translation strings are all under a single parent key in the `views.json` locale file.
 * Then, call this function with the key in i18n notation.
 * The result is a t function which is scoped to the key in question.
 *
 * @param prefix i18n key prefix(es). Pass multiple prefixes to use i18next's built-in fallback mechanism.
 */
export default function useScopedTranslation(...prefixes: string[]) {
  const { t: translate } = useTranslation();

  function t<TResult extends TFunctionResult = string>(
    key: string,
    options?: TOptions<StringMap> | string
  ) {
    const keys = prefixes.map((p) => `views:${p}.${key}`);
    return translate<TResult>(keys, options);
  }

  return { t };
}

// connect-permissions.tsx

interface Props {
  platform: 'google' | 'facebook';
}

interface PlatformStrings {
  heading: string;
  lead: string;
  permissionsList: Array<{
    permission: string;
    reason: string;
  }>;
}

const ReviewsPlatformConnectPermissionsPage: FC<Props> = ({
  platform,
}) => {
  const { t } = useScopedTranslation('reviews.connectPermissions');

  const {
    heading,
    lead,
    permissionsList,
    permissionsNecessaryWarning,
  } = t<PlatformStrings>(platform, { returnObjects: true });

  return (
    // ... JSX including the strings from above
  );
};
```

As you can see, I'm dynamically fetching the correct strings according to the `platform` prop passed in to the component. The strings themselves are in the same structure - some are regular strings, others nested inside an array.

The problem I have when upgrading to 11.8 is that the `<TResult>` generic inside my `use-scoped-translation` hook is no longer valid, as the `t` function from react-i18next expects this to be something else, and sadly I don't really know how to proceed. If you can offer any insight I would be most grateful!

Screenshot 2021-01-04 at 15 53 22

**Comment by jamuhl** [comment]
@fiznool I'm sorry about that. We try to follow semver as good as possible...I should have consult @pedrodurek regarding the version impact before publishing. Will try to improve on that in future.

**Comment by pedrodurek** [comment]
Hey @fiznool, sorry about that, we had to switch the generic types because optional type parameters must always come first, and the `key` parameter is optional. 
To sort that out you have to simply pass `string` as the first argument (key).
```ts
return translate<string, TResult>(keys, options);
```

If you want to type all keys and return type from `t` function, take a look at some examples: https://github.com/i18next/react-i18next/tree/master/example/react-typescript4.1

**Comment by pedrodurek** [comment]
> Hey @joelpoloney, unfortunately that's the only way. That's a typescript behaviour ("deficiency"), when we concatenate strings, typescript automatically infers the type as string, and in order to get the literal values, we must assert the value.
https://github.com/i18next/react-i18next/pull/1193#issuecomment-752283441

Hey @joelpoloney and @DCzajkowski, this is going to fixed on [typescript 4.2](https://devblogs.microsoft.com/typescript/announcing-typescript-4-2-beta/#smarter-type-alias-preservation). Take a look at the example [here](https://www.typescriptlang.org/play?ts=4.2.0-beta#code/MYewdgzgLgBAZgVzMGBeGAKAhgJwOYBcMA5AA5ZQAWAdFCNRCALYCmVAlmHtTixAgCMm7KLREAbFsRgAfEuSq16jVhy4NBw0VAlSAlGgB8MAN4BfAFBXQkWOwgBlTSKgsAJmnhZxEFgG4LG2gYdgBGAA4wACEsX097JyEXdxgAfnkKGjoGZjZKTm5efiSoaSIyTKUc1Xz1Yq1iAMDwYJ0oSU9EZAwAAwASEzDImN8zMXaWHr0moNgWAA8sJlJJULR+wYjo2JYxtskevyA). It's still in beta version, but it's intended to be released on February.

**Comment by joelpoloney** [comment]
Ah, good to know. Thank you for the update!

**Comment by pierpo** [comment]
Thanks for the amazing work on this. It seems to work perfectly well!

I do have an issue concerning passing translations dynamically, though.
I am trying to type a component that accepts a translation key. Is there an exported type that would represent all the possible translation keys given my JSON file?

```tsx
interface PropsType {
  myTranslation: string // can I use an exposed type here???
}

const MyComponent = (props: PropsType) => {
  return <div>t(props.myTranslation)</div>
}
```

**Comment by pedrodurek** [comment]
Hey @pierpo, thanks for your feedback!
`t` function will only accept literal values.
You can rely on the `TFuncKey` coming from `react-i18next` package, but you must pass the namespace as the first argument. Like this:
```ts
interface PropsType {
  myTranslation: TFuncKey<'ns1'> // In case you have only one namespace, just pass 'translation'
}
```

Alternatively, you can pass the actual literal values accepted. Like this:
```ts
interface PropsType {
  myTranslation: 'key.name' | 'key.description';
}
```
Or something like:
```ts
interface PropsType {
  myTranslation: 'name' | 'description';
}
...
{t(`key.${props.myTranslation}` as const)}
```

**Comment by pierpo** [comment]
Thanks for the super quick answer :blush: 
This is exactly what I was looking for!

> ```ts
> interface PropsType {
>   myTranslation: TFuncKey<'ns1'> // In case you have only one namespace, just pass 'translation'
> }
> ```

Thanks again for the amazing work. I've been waiting for this feature for years!

**Comment by ertrzyiks** [comment]
Is there any example out there showing how to use `TInterpolationMap` to also provide type safety for the interpolation params?

I see some potential to provide the mapping type directly to the `t` function,
```
export interface TFunction<N extends Namespace> {
  <K extends TFuncKey<N>, TInterpolationMap extends object = StringMap>
```
and that's the only option currently available, right? Can we expose a similar augmentation way to type the params?

My point is to make sure that `version` is present and is a string in this example:
```
{
  "title": "Welcome to react using react-i18next and typescript {{version}}",
}

<h2>{t('ns1:title', { version: '4.1' })}</h2>
```

**Comment by pedrodurek** [comment]
Hey @ertrzyiks, right now, we don't support type safety for the interpolation params, but that's something that I was considering since the very beginning. The reason for not implementing is because typescript doesn't support getting literal values from JSON files yet https://github.com/microsoft/TypeScript/issues/32063, but I'll keep you posted once it does.

**Comment by andreialecu** [comment]
@pedrodurek I'm also interested in type safe interpolations.

Regarding the issue you linked, there's a workaround for it possible, by transforming the json file into a `.d.ts`. See: https://github.com/microsoft/TypeScript/issues/32063#issuecomment-800162987

Mind explaining how you'd envision it to work? 

It's not clear how `version` would be extracted from a string like: `"Welcome to react using react-i18next and typescript {{version}}",`, which isn't json.

**Pull request merged**